app/module/drivers/sensor/battery/battery_common.c

```diff
@@ -29,15 +29,37 @@ int battery_channel_get(const struct battery_value *value, enum sensor_channel c
     return 0;
 }
 
-uint8_t lithium_ion_mv_to_pct(int16_t bat_mv) {
-    // Simple linear approximation of a battery based off adafruit's discharge graph:
-    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
+uint8_t lithium_ion_mv_to_pct(int16_t batt_mv) {
+    // Lookup table of slope formulas for calculating remaining battery capacity.
+    // The original values used to calculate slopes come from:
+    //
+    //       https://blog.ampow.com/lipo-voltage-chart/
+    //
+    struct lookup_point {
+        float millivolts;
+        float percent;
+    };
 
-    if (bat_mv >= 4200) {
+    static const struct lookup_point battery_lookup[] = {
+        {.millivolts = 4200, .percent = 100}, {.millivolts = 3870, .percent = 60},
+        {.millivolts = 3690, .percent = 10},  {.millivolts = 3610, .percent = 5},
+        {.millivolts = 3270, .percent = 0},
+    };
+
+    if (batt_mv > 4200) {
         return 100;
-    } else if (bat_mv <= 3450) {
-        return 0;
     }
 
-    return bat_mv * 2 / 15 - 459;
-}
\ No newline at end of file
+    for (int i = 1; i < ARRAY_SIZE(battery_lookup); i++) {
+        struct lookup_point one = battery_lookup[i - 1];
+        struct lookup_point two = battery_lookup[i];
+        if (batt_mv >= two.millivolts) {
+            float slope = (one.percent - two.percent) / (one.millivolts - two.millivolts);
+            float offset = one.percent - slope * one.millivolts;
+            float batt_percent = (slope * batt_mv) + offset;
```

Comment:
This can be simplified a bit to use only one multiply and one divide. If the divide is placed last, then it should also work using only integer math, so there's no need to cast anything to a float anywhere.

The formula for linear interpolation between two points (x0, y0) and (x1, y1) is

```c
y = y0 + (y1 - y0) * (x - x0) / (x1 - x0)
```

which using the variable names here is

```c
return one.percent + (two.percent - one.percent) * (batt_mv - one.millivolts) / (two.millivolts - one.millivolts);
```

or pulling some of that out into separate variables to make it easier to read

```c
const int t = batt_mv - one.millivolts;
const int dx = two.millivolts - one.millivolts;
const int dy = two.percent - one.percent;

return one.percent + dy * t / dx;
```

(note that because you ordered the table in descending order, many of the values here would end up negative, but they will all cancel out)

Reply:
Hmmm, I should have looked up the standard form for linear interpolation!

I just did what I see the numbers doing, this is much much nicer!

More generally you've spent more time nursing this PR than it would have taken you to do it yourself, I really appreciate that. I've learned a little about C++ as a result.